Sources/StatsD.swift

```diff
@@ -18,30 +19,69 @@ protocol StatsdProtocol {
     func gauge(_ bucket: String, delta: Int)
 }
 
-public class StatsD: StatsdProtocol {
+public class StatsD: NSObject, StatsdProtocol {
+
     let transport: Transport
+    private let storage: MemoryStorage<Metric>
+    private var flush: Flushable
+
+    public convenience init(transport: Transport) {
+        let storage = MemoryStorage<Metric>()
+        let schemes: [FlushScheme] = [AppCycleFlushScheme(),
+                                      IntervalFlushScheme(),
+                                      ]
+        let flush = Flush(schemes: schemes)
+        self.init(transport: transport, storage: storage, flush: flush)
+    }
 
-    public init(transport: Transport) {
+    init(transport: Transport, storage: MemoryStorage<Metric>, flush: Flushable) {
         self.transport = transport
+        self.flush = flush
+        self.storage = storage
+        super.init()
+
+        self.flush.delegate = self
+        flush.start()
     }
 
     public func increment(_ bucket: String, by value: Int = 1) {
-        transport.write(data: Counting(name: bucket, value: "\(value)").metricData, completion: nil)
+        storage.set(item: Counting(name: bucket, value: "\(value)"), forKey: bucket)
     }
 
     public func set(_ bucket: String, value: String) {
-        transport.write(data: Sets(name: bucket, value: value).metricData, completion: nil)
+        storage.set(item: Sets(name: bucket, value: value), forKey: bucket)
     }
 
     public func timing(_ bucket: String, value: Int) {
-        transport.write(data: Timing(name: bucket, value: "\(value)").metricData, completion: nil)
+        storage.set(item: Timing(name: bucket, value: value), forKey: bucket)
     }
 
     public func gauge(_ bucket: String, value: UInt) {
-        transport.write(data: Gauge(name: bucket, value: value).metricData, completion: nil)
+        storage.set(item: Gauge(name: bucket, value: value), forKey: bucket)
     }
 
     public func gauge(_ bucket: String, delta: Int) {
-        transport.write(data: Gauge(name: bucket, delta: delta).metricData, completion: nil)
+        storage.set(item: Gauge(name: bucket, delta: delta), forKey: bucket)
+    }
+}
+
+extension StatsD: FlushDelegate {
+
+    func flush(_ sender: Flushable) {
+        guard !storage.isEmpty else {
+            return
+        }
+
+        let batch = mapToBatch()
+        transport.write(data: batch) {[unowned self] (error) in
+            guard error == nil else {
+                return
+            }
+            self.storage.removeAll()
+        }
+    }
+
+    private func mapToBatch() -> String {
```

Comment:
`mapToBatch` is a terrible name, I'd suggest create a static function inside StatsD. Something like
```
static func accumulate(metric: [Metric]) -> String  {

}
```
Later on maybe we'll even have different Accumulators for different metrics.